### PR TITLE
fix #7050: add mode to disable users

### LINKED ISSUE #7050: Add mode to disable users
Should also come with a cli to do this. This would not delete the user, but simply prevent the user from accessing anything once logged in.

### PR DESCRIPTION
**Pull request by rorour**
**contains db upgrade**

### CODE REVIEW

File: sirepo/api_auth.py
Comment (robnagler) [inline sirepo/api_auth.py:46]: REQUIRE_DISABLED will not be an expected by an API. It's a role that is checked in a way similar to the sym_type roles. It's more of an attribute-based access than a role.  It needs to be checked always, whenever a user is required so it needs to be added to require_user whenever the user is logged in.

File: sirepo/api_perm.py
Comment (robnagler) [inline sirepo/api_perm.py:32]: Not necessary. Sorry for steering you incorrectly.

File: sirepo/auth_role.py
Comment (robnagler) [inline sirepo/auth_role.py:13]: Let's call this ROLE_ACCOUNT_DISABLED so it's a bit clearer what's disabled.

File: sirepo/server.py
Comment (robnagler) [inline sirepo/server.py:477]: This would be insufficient in any event if they bookmarked a simulation. This security has to be pervasive.

File: tests/disabled_user_test.py
Comment (robnagler) [inline tests/disabled_user_test.py:16]: This is actually interesting. We don't necessarily want to return this particular role in auth_state, because it gives an attacker a reason for the forbidden, a way to probe the system. Therefore, auth_state will have to filter this out.

File: tests/disabled_user_test.py
Comment (robnagler) [inline tests/disabled_user_test.py:22]: This should test a few APIs, e.g. running a simulation. You'll need to cache the request info for this case so that the request goes straight to the uri.

**Review by robnagler — CHANGES_REQUESTED** [review]

File: tests/disabled_user_test.py
Comment (rorour) [reply tests/disabled_user_test.py:16]: With our code being open source and this PR being public it seems like they could figure it out regardless

Comment (rorour) [reply tests/disabled_user_test.py:16]: I understand that, but someone could still look through our code/conversations and suspect that they have this role. I don't see how filtering out this role from auth_state makes things any more secure.

Comment (robnagler) [reply tests/disabled_user_test.py:16]: Here's a new thought: when two programmers disagree, there's something wrong with the design. @rorour, you discovered a problem. The original solution to use "negative role" is an incorrect hack. What we would do in a password-based systems is disable the password so they couldn't login any more. Even that does not give a reason or have any audit support. I chose that solution because we have cli's that support this already. Maybe when two programmers disagree, the customer is being lazy (or cheap :). The right solution is more expensive, or is it?

With moderation, we deny and have a person bound to it with a time. This is a similar problem, but it's different. If we were doing true RBAC, we would have a permission to use the system at all. We could have a db upgrade that adds a role "user" or something like that. require_user would check this role, because that's the surrogate. (In BOP, we had such a role as all access, even [visitors got a role and permissions](https://github.com/biviosoftware/perl-Bivio/blob/master/Util/SQL.pm#L962).) 

We could also add then a UserRoleInvite record when we deny the user with a simple command line utility. The login in require_user would be the same, but it would be positive: in order to validate as a user there would have to be a role. auth.login wouldn't allow the user to login either. Perhaps that's giving information that they are denied, but as you note, they will be able to detect when they are denied in any event, and infer something. The key is that we don't return authState with roles unless they are logged in so they don't really get any more information with auth_state.

Thoughts?

Comment (rorour) [reply tests/disabled_user_test.py:16]: Yes, I think making the role positive rather than negative is a great improvement on this idea.

Would new users have to wait for approval?

Comment (robnagler) [reply tests/disabled_user_test.py:16]: > Would new users have to wait for approval?

We won't build this yet, but this might be an option. For now, all users will get the role on creation.

**Comment by rorour** [comment]
(code is not ready for re-review)
I think just calling the role "user" is a little confusing, at least it makes function names like "check_user_role" ambiguous. Maybe one of these would be better:
- enabled_user
- allowed_user 
- account_enabled

**Comment by robnagler** [comment]
One of the problems is that we are abusing the notion of "role". User, Visitor, Admin, etc. are roles. sim_type_flash is not a role, it's a capability or permission. Here's a [nice explanation](https://wpmarmite.com/en/glossary/user-role/).

We could add a keyword, e.g. role_user or user_role. The actual role would be "user". Technically, we should have a role called "visitor". The check for cookieless_user is really weird, of course. Another abuse. 

One possibility is registered_user. However this implies there are unregistered users. Maybe an unregistered_user is a visitor.

I think this is an opportunity to reorient ourselves around roles. We can't fix it, but we can at least understand that we are abusing roles and accept true role names.

**Comment by rorour** [comment]
> We could also add then a UserRoleInvite record when we deny the user with a simple command line utility. 

Can you explain more of what you had in mind for this CLI? I was thinking of adding a function `disable_user(uid_or_email)` which handles the UserRoleInvite entry to pkcli/roles, but how would we get the moderator uid?

**Comment by robnagler** [comment]
> how would we get the moderator uid?

An argument to the cli function.

**Comment by rorour** [comment]
What do you think of adding a new table for this? 
UserRoleInvite requires a not null value for column token (part of the moderation request process), and making that nullable requires modifying the table schema [which is not trivial in sqlite](https://stackoverflow.com/questions/4007014/alter-column-in-sqlite).
There is also no "invite" involved in this case, so that suggests this table isn't right for this purpose.

**Comment by rorour** [comment]
@e-carlin and I just had a discussion. Is "token" really adding any security benefit to role moderation?

Maybe we could drop token from UserRoleInvite and still use it for this purpose. I still think a rename would be good in that case.

**Comment by robnagler** [comment]
> Is "token" really adding any security benefit to role moderation?

It wasn't put in there for security afaict. It is used to find the record. It's not used that way in the moderate list when clicking on one of the links (approve, deny, clarify). It's a primary key, not a token so it's misnamed. The token could be switched to (uid, role), which is the actual primary key.

**Comment by rorour** [comment]
@robnagler ping

**Comment by robnagler** [comment]
ya, working through them...

File: sirepo/db_upgrade.py
Comment (robnagler) [inline sirepo/db_upgrade.py:93]: Ident this so easier to read. SQL doesn't care about leading whitespace.

File: sirepo/db_upgrade.py
Comment (robnagler) [inline sirepo/db_upgrade.py:99]: Since there's direct SQL above. I think it would better and simpler to write some SQL. There is an efficiency and ACID reason to do this. The shorter the transaction, the better.

```sql
INSERT INTO user_role_t (uid, role, expiration) 
VALUES ((select uid from user_registration_t), "user", NULL);
```

"user" could be an injected string, because it is a constant (no chance of sql injection attacks. This is an upgrade, and denormalizing the value makes it simpler to read and avoids the question of sql injection attack.

File: sirepo/pkcli/roles.py
Comment (robnagler) [inline sirepo/pkcli/roles.py:49]: This is a bit tricky. I would say put the uid_or_email first (like the other functions in roles). This was related to a discussion we had about C++ arguments in a class switching in an API and causing @moellep confusion. There's also the general idea that the first argument is what is the subject and the other args are modifiers.

The second argument is just `moderator` (not moderator uid). It would be convenient for admins to not have to remember their uid. extract the uid lookup code in _parse_args so that it can be called separately.

File: sirepo/pkcli/roles.py
Comment (robnagler) [inline sirepo/pkcli/roles.py:51]: Formatting isn't right. Needs to be indented.

File: sirepo/pkcli/roles.py
Comment (robnagler) [inline sirepo/pkcli/roles.py:84]: This is a Python anachronism. `roles` is statically bound to a specific `[]` instance. Since `[]` is mutable, it is dangerous to do that. You could use `None` or `()` here.

File: tests/adm_and_own_jobs_test.py
Comment (robnagler) [inline tests/adm_and_own_jobs_test.py:44]: remove blank

File: tests/adm_and_own_jobs_test.py
Comment (robnagler) [inline tests/adm_and_own_jobs_test.py:43]: The delete_all_for_column_by_values is incorrect. We want to only delete the admin role.

File: tests/adm_and_own_jobs_test.py
Comment (robnagler) [inline tests/adm_and_own_jobs_test.py:124]: This should be ROLE_ADM. The test doesn't make sense the way it was written or with ROLE_USER. The check is for all ROLE_ADM rows.

File: tests/auth_role_moderation_test.py
Comment (robnagler) [inline tests/auth_role_moderation_test.py:29]: There's a pattern here. If you find yourself exposing assumptions in code, e.g. "all users must have ROLE_USER", then something is wrong. The delete_all calls (in all cases) was from "lazy" coding. Since tests always create new databases, this works, but it makes an assumption about the roles table that it shouldn't.

In this case, what's wanted is to delete the sym_type role and moderation record for myapp for this particular user.

File: tests/pkcli/admin_test.py
Comment (robnagler) [inline tests/pkcli/admin_test.py:36]: _is_empty_table is the incorrect test. Check for this user in the table(s). The assumption that there is only one user in the db is unnecessary.

File: tests/pkcli/roles_test.py
Comment (robnagler) [inline tests/pkcli/roles_test.py:59]: Again this makes an incorrect assumption. The test should only make sure that there's an entry for sim_type flash.

File: tests/role_user_test.py
Comment (robnagler) [inline tests/role_user_test.py:10]: test_disable_user.

File: tests/role_user_test.py
Comment (robnagler) [inline tests/role_user_test.py:1]: delete

File: tests/role_user_test.py
Comment (robnagler) [inline tests/role_user_test.py:2]: This tests pkcli.admin.disable_user. Make pkcli.admin2_test. pkcli.admin_test does too much to the db and doesn't use `fc` (that is it doesn't start a server).

File: tests/supervisor_purge_free_sims_test.py
Comment (robnagler) [inline tests/supervisor_purge_free_sims_test.py:56]: Excellent!

**Review by robnagler — CHANGES_REQUESTED** [review]

File: sirepo/auth/__init__.py
Comment (rorour) [reply sirepo/auth/__init__.py]: [793c0b5](https://github.com/radiasoft/sirepo/pull/7070/commits/793c0b547ac000e96d88c0a70af24f34c7ad011f)

File: sirepo/auth/__init__.py
Comment (rorour) [reply sirepo/auth/__init__.py]: [793c0b5](https://github.com/radiasoft/sirepo/pull/7070/commits/793c0b547ac000e96d88c0a70af24f34c7ad011f)

File: sirepo/db_upgrade.py
Comment (rorour) [reply sirepo/db_upgrade.py]: [793c0b5](https://github.com/radiasoft/sirepo/pull/7070/commits/793c0b547ac000e96d88c0a70af24f34c7ad011f)

File: tests/adm_and_own_jobs_test.py
Comment (rorour) [reply tests/adm_and_own_jobs_test.py]: [793c0b5](https://github.com/radiasoft/sirepo/pull/7070/commits/793c0b547ac000e96d88c0a70af24f34c7ad011f)

File: sirepo/auth/__init__.py
Comment (rorour) [reply sirepo/auth/__init__.py:531]: [609c03c](https://github.com/radiasoft/sirepo/pull/7070/commits/609c03c32717dcd27067fe26bc7c2bbb4ee16dc1)

File: sirepo/pkcli/roles.py
Comment (rorour) [reply sirepo/pkcli/roles.py:50]: [609c03c](https://github.com/radiasoft/sirepo/pull/7070/commits/609c03c32717dcd27067fe26bc7c2bbb4ee16dc1)

File: sirepo/pkcli/roles.py
Comment (rorour) [reply sirepo/pkcli/roles.py]: [609c03c](https://github.com/radiasoft/sirepo/pull/7070/commits/609c03c32717dcd27067fe26bc7c2bbb4ee16dc1)

File: tests/role_user_test.py
Comment (rorour) [reply tests/role_user_test.py]: [609c03c](https://github.com/radiasoft/sirepo/pull/7070/commits/609c03c32717dcd27067fe26bc7c2bbb4ee16dc1)

File: tests/role_user_test.py
Comment (rorour) [reply tests/role_user_test.py]: [609c03c](https://github.com/radiasoft/sirepo/pull/7070/commits/609c03c32717dcd27067fe26bc7c2bbb4ee16dc1)

File: tests/role_user_test.py
Comment (rorour) [reply tests/role_user_test.py]: [609c03c](https://github.com/radiasoft/sirepo/pull/7070/commits/609c03c32717dcd27067fe26bc7c2bbb4ee16dc1)

File: sirepo/auth/__init__.py
Comment (rorour) [reply sirepo/auth/__init__.py:705]: I'm having a hard time understanding the second half of your comment. Is this what you mean?

```python3
    def _create_user(self, module, want_login):
        def _create_roles_for_new_user(self, method, uid):
            r = sirepo.auth_role.for_new_user(is_guest=method == METHOD_GUEST)
            if r:
                self.qcall.auth_db.model("UserRole").add_roles(roles=r, uid=uid)

        u = simulation_db.user_create()
        if want_login:
            self._login_user(module, u)
        self._create_roles_for_new_user(module.AUTH_METHOD, uid=u)
        return u
```

Comment (robnagler) [reply sirepo/auth/__init__.py:705]:
```py
def _create_user(self, module, want_login):
     u = simulation_db.user_create()
     r = sirepo.auth_role.for_new_user(is_guest=method == METHOD_GUEST)
     if r:
         self.qcall.auth_db.model("UserRole").add_roles(roles=r, uid=uid)
     if want_login:
         self._login_user(module, u)
     return u
```

The want_login has to go at the end because there's check_role_user which needs to be added above.

Comment (rorour) [reply sirepo/auth/__init__.py]: [634efd8](https://github.com/radiasoft/sirepo/pull/7070/commits/634efd840a1338c81fced52736fe4cfde315bb63)

File: sirepo/auth/__init__.py
Comment (rorour) [reply sirepo/auth/__init__.py]: [634efd8](https://github.com/radiasoft/sirepo/pull/7070/commits/634efd840a1338c81fced52736fe4cfde315bb63)

File: sirepo/auth_role_moderation.py
Comment (rorour) [reply sirepo/auth_role_moderation.py]: [634efd8](https://github.com/radiasoft/sirepo/pull/7070/commits/634efd840a1338c81fced52736fe4cfde315bb63)

File: sirepo/auth_role_moderation.py
Comment (rorour) [reply sirepo/auth_role_moderation.py]: [634efd8](https://github.com/radiasoft/sirepo/pull/7070/commits/634efd840a1338c81fced52736fe4cfde315bb63)

File: sirepo/auth_role_moderation.py
Comment (rorour) [reply sirepo/auth_role_moderation.py]: [634efd8](https://github.com/radiasoft/sirepo/pull/7070/commits/634efd840a1338c81fced52736fe4cfde315bb63)

File: sirepo/db_upgrade.py
Comment (rorour) [reply sirepo/db_upgrade.py:103]: [634efd8](https://github.com/radiasoft/sirepo/pull/7070/commits/634efd840a1338c81fced52736fe4cfde315bb63)

File: sirepo/pkcli/roles.py
Comment (rorour) [reply sirepo/pkcli/roles.py]: [634efd8](https://github.com/radiasoft/sirepo/pull/7070/commits/634efd840a1338c81fced52736fe4cfde315bb63)

File: tests/adm_and_own_jobs_test.py
Comment (rorour) [reply tests/adm_and_own_jobs_test.py]: [4508278](https://github.com/radiasoft/sirepo/pull/7070/commits/45082783727593c14d65c391f7a7e8c66412753f)

File: sirepo/db_upgrade.py
Comment (robnagler) [inline sirepo/db_upgrade.py]: Does this work? I think ROLE_USER needs to be quoted:

`f"VALUES ((select uid from user_registration_t), '{sirepo.auth_role.ROLE_USER}', NULL);"`

**Review by robnagler — CHANGES_REQUESTED** [review]
One fix required.

File: sirepo/db_upgrade.py
Comment (rorour) [reply sirepo/db_upgrade.py]: That is an old commit.

**Pull request merged**
